### PR TITLE
Fix redirect_uri resolver

### PR DESCRIPTION
**Pull request by leshalv**
Fixes gh-1012

### CODE REVIEW

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java [hunk 1]
```diff
@@ -124,17 +124,26 @@ public Authentication authenticate(Authentication authentication) throws Authent
 					authorizationCodeRequestAuthentication, registeredClient);
 		}
 
+		OAuth2AuthorizationRequest authorizationRequest = OAuth2AuthorizationRequest.authorizationCode()
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java:127]: Please revert all changes and instead add parameter `OAuth2AuthorizationCodeRequestAuthenticationToken authorizationCodeRequestAuthentication` to `resolveRedirectUri()` allowing usage of `authorizationCodeRequestAuthentication.redirectUri`.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thank you for the PR @leshalv.

Please see the review comment and also please add a test to confirm the fix.

**Comment by leshalv** [comment]
Ok, I've made the modification.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java [hunk 2]
```diff
@@ -333,7 +333,7 @@ private static void throwError(OAuth2Error error, String parameterName,
 			OAuth2AuthorizationCodeRequestAuthenticationToken authorizationCodeRequestAuthentication,
 			RegisteredClient registeredClient, OAuth2AuthorizationRequest authorizationRequest) {
 
-		String redirectUri = resolveRedirectUri(authorizationRequest, registeredClient);
+		String redirectUri = resolveRedirectUri(authorizationCodeRequestAuthentication,authorizationRequest, registeredClient);
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java:336]: Please ensure spacing between parameter values.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java [hunk 3]
```diff
@@ -350,7 +350,10 @@ private static void throwError(OAuth2Error error, String parameterName,
 		throw new OAuth2AuthorizationCodeRequestAuthenticationException(error, authorizationCodeRequestAuthenticationResult);
 	}
 
-	private static String resolveRedirectUri(OAuth2AuthorizationRequest authorizationRequest, RegisteredClient registeredClient) {
+	private static String resolveRedirectUri(OAuth2AuthorizationCodeRequestAuthenticationToken authorizationCodeRequestAuthentication,OAuth2AuthorizationRequest authorizationRequest, RegisteredClient registeredClient) {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java:353]: Please ensure spacing between parameter values.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java [hunk 4]
```diff
@@ -350,7 +350,10 @@ private static void throwError(OAuth2Error error, String parameterName,
...
+		if (authorizationCodeRequestAuthentication!=null && StringUtils.hasText(authorizationCodeRequestAuthentication.getRedirectUri())){
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeRequestAuthenticationProvider.java:354]: Please ensure proper spacing.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the updates @leshalv. Please see review comment for minor formatting changes.

Also, before we merge this, we need a test the proves the fix has been applied. You should be able to modify the existing test `OAuth2AuthorizationCodeRequestAuthenticationProviderTests.authenticateWhenPkceRequiredAndMissingCodeChallengeThenThrowOAuth2AuthorizationCodeRequestAuthenticationException()` to demonstrate the fix was applied. Thanks.

**Comment by jgrandja** [comment]
@leshalv As mentioned in this [comment](https://github.com/spring-projects/spring-authorization-server/pull/1013#pullrequestreview-1250596906), a test needs to be added as part of this PR. Are you able to add it?

**Comment by leshalv** [comment]
> @leshalv As mentioned in this [comment](https://github.com/spring-projects/spring-authorization-server/pull/1013#pullrequestreview-1250596906), a test needs to be added as part of this PR. Are you able to add it?

Can you help me improve it?

**Comment by jgrandja** [comment]
@leshalv 

> Can you help me improve it?

Do you need help with writing the test? Do you want me to write it and then merge?

**Comment by leshalv** [comment]
> 

yes

**Comment by leshalv** [comment]
> @leshalv
> 
> > Can you help me improve it?
> 
> Do you need help with writing the test? Do you want me to write it and then merge?

yes

**Comment by jgrandja** [comment]
@leshalv Thanks again for reporting this. I added a polish commit with tests and is now merged.

**Pull request closed without merging**